File: src/main/java/com/ada/shared/observability/ContextMetricsAspect.java
package com.ada.shared.observability;

import io.micrometer.core.instrument.*; import org.aspectj.lang.*; import org.aspectj.lang.annotation.*; import org.springframework.stereotype.Component;
@Aspect @Component public class ContextMetricsAspect { private final MeterRegistry registry; public ContextMetricsAspect(MeterRegistry r){registry=r;} @Around("@within(measuredContextItem)") public Object measure(ProceedingJoinPoint jp,MeasuredContextItem a)throws Throwable{var s=Timer.start(registry);registry.counter("ada_context_item_invocations_total","component",a.value()).increment();try{return jp.proceed();}finally{s.stop(registry.timer("ada_context_item_duration_seconds","component",a.value()));}} }

File: src/main/kotlin/com/ada/shared/observability/ContextMetricsAspect.kt
package com.ada.shared.observability

import io.micrometer.core.instrument.MeterRegistry
import io.micrometer.core.instrument.Timer
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.springframework.stereotype.Component

@Aspect
@Component
class ContextMetricsAspect(
    private val registry: MeterRegistry,
) {
    @Around("@within(measuredContextItem)")
    fun measure(joinPoint: ProceedingJoinPoint, measuredContextItem: MeasuredContextItem): Any? {
        val sample = Timer.start(registry)
        val tags = arrayOf("component", measuredContextItem.component)
        registry.counter("ada_context_item_invocations_total", *tags).increment()
        return try {
            joinPoint.proceed()
        } finally {
            sample.stop(registry.timer("ada_context_item_duration_seconds", *tags))
        }
    }
}
